bin/varnishd/cache/cache_vrt_vcl.c

```diff
@@ -517,3 +517,51 @@ VCL_##func##_method(struct vcl *vcl, struct worker *wrk,		\
 }
 
 #include "tbl/vcl_returns.h"
+
+/*--------------------------------------------------------------------
+ */
+
+VCL_STRING
+VRT_check_call(VRT_CTX, VCL_SUB sub)
+{
+	VCL_STRING err = NULL;
+	enum vcl_func_fail_e fail;
+
+	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);
+	CHECK_OBJ_NOTNULL(sub, VCL_SUB_MAGIC);
+
+	AN(sub->func);
+	sub->func(ctx, VSUB_CHECK, &fail);
+
+	switch (fail) {
+	case VSUB_E_OK:
+		break;
+	case VSUB_E_METHOD:
+		err = WS_Printf(ctx->ws, "Dynamic call to \"sub %s{}\""
+		    " not allowed from here", sub->name);
+		if (err == NULL)
+			err = "Dynamic call not allowed and workspace overflow";
+		break;
+	case VSUB_E_RECURSE:
+		err = WS_Printf(ctx->ws, "Recursive dynamic call to"
+		    " \"sub %s{}\"", sub->name);
+		if (err == NULL)
+			err = "Recursive dynamic call and workspace overflow";
+		break;
+	default:
+		INCOMPL();
+	}
+
+	return (err);
+}
+
+VCL_VOID
+VRT_call(VRT_CTX, VCL_SUB sub)
+{
+
+	CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);
+	CHECK_OBJ_NOTNULL(sub, VCL_SUB_MAGIC);
+
+	AN(sub->func);
+	sub->func(ctx, VSUB_DYNAMIC, NULL);
```

Comment:
Should we check the context handling to avoid making progress somewhere:tm: after a return action?

Reply:
see 1:1 IRC: I do not understand what you mean.